dotcom-rendering/src/lib/assets.ts

```diff
@@ -47,31 +37,63 @@ const isDev = process.env.NODE_ENV === 'development';
 
 export const ASSET_ORIGIN = decideAssetOrigin(process.env.GU_STAGE, isDev);
 
-export const getScriptArrayFromFile = (
+const getManifest = (path: string): AssetHash => {
+	try {
+		const assetHash: unknown = JSON.parse(
+			readFileSync(resolve(__dirname, path), { encoding: 'utf-8' }),
+		);
+		if (typeof assetHash != 'object' || assetHash === null)
+			throw new Error('Not a valid AssetHash type');
+
+		/** @TODO validate the object */
+		return assetHash as AssetHash;
+	} catch (e) {
+		console.error('Could not load manifest in: ', path);
+		console.error('Some filename lookups will fail');
+		return {};
+	}
+};
+
+export type ManifestPath = `./manifest.${string}.json`;
+
+const getScripts = (
+	manifestPaths: Array<ManifestPath>,
 	file: `${string}.js`,
-): { src: string; legacy?: boolean }[] => {
+): string[] => {
 	if (!file.endsWith('.js'))
 		throw new Error('Invalid filename: extension must be .js');
 
-	const filename = isDev ? file : manifest[file];
-	const legacyFilename = isDev
-		? file.replace('.js', '.legacy.js')
-		: legacyManifest[file];
-
-	const scripts = [];
-
-	if (filename) {
-		scripts.push({
-			src: `${ASSET_ORIGIN}assets/${filename}`,
-			legacy: false,
-		});
-	}
-	if (legacyFilename) {
-		scripts.push({
-			src: `${ASSET_ORIGIN}assets/${legacyFilename}`,
-			legacy: true,
-		});
+	if (isDev) {
+		return [
+			`${ASSET_ORIGIN}assets/${file.replace('.js', '.modern.js')}`,
+			`${ASSET_ORIGIN}assets/${file.replace('.js', '.legacy.js')}`,
+		];
 	}
 
-	return scripts;
+	return manifestPaths.map((manifestPath) => {
+		const manifest = getManifest(manifestPath);
+		const filename = manifest[file];
+
+		return `${ASSET_ORIGIN}assets/${filename}`;
+	});
 };
+
+export const getScriptsFromManifest =
+	(manifestPaths: Array<`./manifest.${string}.json`>) =>
+	(file: `${string}.js`): ReturnType<typeof getScripts> =>
+		getScripts(manifestPaths, file);
+
+export const generateScriptTags = (scripts: Array<string | false>): string[] =>
+	scripts.filter(isString).map((script) => {
+		if (script.includes('.legacy.')) {
+			return `<script defer nomodule src="${script}"></script>`;
+		}
+		if (script.includes('.modern.')) {
+			return `<script type="module" src="${script}"></script>`;
+		}
```

Comment:
This is a lot simpler, but assumes that “external” does not contain these patterns.

Reply:
Could be a little bit safer `.endsWidth('.modern.js')`?

Reply:
Not quite, because of the fact that we might have a hash. We’ve implemented a regular expression matching in ec3c5e6aef1bb8f78ce33b4cf7bcf3a441f133b9.